contracts/staking/StakingBase.sol

```diff
@@ -661,15 +894,47 @@ abstract contract StakingBase is ERC721TokenReceiver {
                 lastAvailableRewards += reward;
                 availableRewards = lastAvailableRewards;
             } else {
-                _withdraw(multisig, reward);
+                // Get reward receivers and amounts based on the distribution type
+                (address[] memory receivers, uint256[] memory amounts) =
+                    _getRewardReceiversAndAmounts(serviceId, sInfo.owner, sInfo.multisig, sInfo.reward, sInfo.rewardDistributionInfo);
+
+                // Transfer reward amounts to specified receivers
+                _withdraw(receivers, amounts);
+
+                emit RewardClaimed(epochCounter, serviceId, msg.sender, sInfo.multisig, sInfo.nonces, receivers, amounts);
             }
         }
 
-        if (enforced) {
-            emit ServiceForceUnstaked(epochCounter, serviceId, msg.sender, multisig, nonces, reward, lastAvailableRewards);
-        } else {
-            emit ServiceUnstaked(epochCounter, serviceId, msg.sender, multisig, nonces, reward, lastAvailableRewards);
+        // Transfer the service back to the owner
+        // Note that the reentrancy is not possible due to the ServiceInfo struct being deleted
+        IService(serviceRegistry).transferFrom(address(this), msg.sender, serviceId);
+
+        emit ServiceUnstaked(epochCounter, serviceId, msg.sender, sInfo.multisig, sInfo.nonces, lastAvailableRewards, enforced);
+    }
+
+    /// @dev Withdraws reward amounts.
+    /// @notice The balance is always greater or equal the sum of amounts, as follows from the contract logic.
+    /// @param receivers Set of receiver addresses.
+    /// @param amounts Corresponding amounts to withdraw.
+    function _withdraw(address[] memory receivers, uint256[] memory amounts) internal virtual {
+        // Get current balance
+        uint256 updatedBalance = balance;
+
+        // Note that if array lengths are different, the withdraw reverts, but that's a responsibility of service owner
```

Comment:
this will just revert whether amounts.length > receivers.length, right? 
Should we add an explicit check for the len?

Reply:
Corrected that one, inserted a check for custom withdraw